src/utils/time.js

```diff
@@ -6,3 +10,54 @@ export function isValidDateString (value) {
     return false
   }
 }
+
+const mapQuasarTimeUnitToStelace = {
+  m: 'minutes',
+  h: 'hours',
+  d: 'days',
+  M: 'months'
+}
+
+function getQuasarTimeUnit (timeUnit) {
+  return mapQuasarTimeUnitToStelace[timeUnit]
+}
+
+export function convertEndDateFromAPIToUI (endDate, { startDate, timeUnit = 'd', includeEndDate = true } = {}) {
+  if (!includeEndDate) return endDate
+
+  const quasarTimeUnit = getQuasarTimeUnit(timeUnit)
+
+  if (!endDate) return endDate
+  const newEndDate = addToDate(new Date(endDate).toISOString(), { [quasarTimeUnit]: -1 }).toISOString()
+
+  // align the offset to start date to avoid non-integer duration
+  const differenceOffset = getOffsetDifference(startDate, newEndDate)
+  if (differenceOffset) {
+    return addToDate(newEndDate, { minutes: differenceOffset }).toISOString()
+  } else {
+    return newEndDate
+  }
+}
+
+export function convertEndDateFromUIToAPI (endDate, { startDate, timeUnit = 'd', includeEndDate = true } = {}) {
+  if (!includeEndDate) return endDate
+
+  const quasarTimeUnit = getQuasarTimeUnit(timeUnit)
+
+  if (!endDate) return endDate
+  const newEndDate = addToDate(new Date(endDate).toISOString(), { [quasarTimeUnit]: 1 }).toISOString()
+
+  // align the offset to start date to avoid non-integer duration
+  const differenceOffset = getOffsetDifference(startDate, newEndDate)
+  if (differenceOffset) {
+    return addToDate(newEndDate, { minutes: -differenceOffset }).toISOString()
+  } else {
+    return newEndDate
+  }
+}
+
+export function getOffsetDifference (startDate, endDate) {
```

Comment:
As discussed on slack, you may want to comment about timezone issues, and `Date.UTC` could be useful too.

Reply:
**Problem description**

The problem I encountered when selecting `startDate` and `endDate` is that the duration isn't an integer for some specific periods. It happens when `startDate` and `endDate` aren't on the same offset due to daylight saving.

For instance when selecting the period `2019/10/27` and `2019/10/28`, current [`DateRangePicker`](https://github.com/stelace/marketplace-demo/blob/dev/src/components/DateRangePicker.vue) returns `2019-10-26T22:00:00.000Z` and `2019-10-28T23:00:00.000Z` in a French system.
The transaction previewed duration is 2.042 days instead of 2 days due to this additional hour.

**Resolution**

Your suggestion to use `Date.UTC` is a good one as we don't need to handle time units inferior than days in calendar.
But I see a different way to solve this problem. In fact the [`DatePickerInput`](https://github.com/stelace/marketplace-demo/blob/fbc4503ae3fbd09d2e1dbe11fc0b5217f709674d/src/components/DatePickerInput.vue) used by `DateRangePicker` directly parse the Quasar date via `new Date().toISOString()`.
https://github.com/stelace/marketplace-demo/blob/fbc4503ae3fbd09d2e1dbe11fc0b5217f709674d/src/components/DatePickerInput.vue#L106-L109

And this leads to the problem mentioned above:
`new Date('2019/10/27') === '2019-10-26T22:00:00.000Z'` (`2019/10/27` is Quasar date)
while
`new Date('2019-10-27') === '2019-10-27T00:00:00.000Z'` (with dashes instead of slashes)

I'll fix this and probably remove the `convertEndDateFromUIToAPI` and `convertEndDateFromAPIToUI` functions as we may not need them anymore.

**Dealing with timezones (later)**

Currently, we're dealing only with plain dates without hours so working with UTC does the job.
However if hours or minutes are introduced, we'll have timezone and offset issues. And this is pretty complex as explained in [this article](https://medium.com/@toastui/handling-time-zone-in-javascript-547e67aa842d). Here's [another article](https://medium.com/@vivekmadurai/how-to-deal-with-date-and-time-across-time-zones-39b1bd747f35) on how deal with date across timezones.

If the platform is specifically intended to work within a timezone, we may add an environment variable specifying it so even if a user from a different timezone views the platform, we can adapt the date display to her with a library like [moment.js](https://momentjs.com).

And if the platform is international, we may add user timezone at signup because the booking time is generally expressed in the owner local time (like Airbnb's bookings).

Reply:
Thanks for these details.

> 2019/10/27 is Quasar date

I doubt it’s [quasar date](https://quasar.dev/quasar-utils/date-utils) default format per se.

It seems you initialley introduced this (European) format in codebase for UI. So deeper refactoring may be appropriate, with formatting applying _after_ logic.

Reply:
After digging the documentation, `QDate` uses the [mask](https://quasar.dev/vue-components/date#Model-mask) `YYYY/MM/DD` by default, that's why the dates returned use this format.
I change this format to `YYYY-MM-DD` so we don't need date string manipulation.

Reply:
Thanks.